### PR TITLE
Feature/gcb 64 integrate with google calendar api

### PR DESCRIPTION
**Pull request by michallaskowski**

### CODE REVIEW

File: GrandCentralBoard/Utilities/GoogleAPIDataProvider.swift [hunk 1]
```diff
@@ -11,10 +11,15 @@ import Operations
 
 enum APIDataError : ErrorType {
     case AuthorizationError
+    case ModelDecodeError(ErrorType)
     case UnderlyingError(NSError)
 }
 
-final class GoogleAPIDataProvider {
+protocol APIDataProvider {
```
Comment (ochococo) [inline GrandCentralBoard/Utilities/GoogleAPIDataProvider.swift]: `APIDataFetching` / `APIDataProviding` - describe ability. 

`APIDataProvider` is better name for a class.

File: GrandCentralBoard/Utilities/GoogleAPIDataProvider.swift [hunk 2]
```diff
@@ -11,10 +11,15 @@ import Operations
...
+    func request(method: Method, url: NSURL, parameters: [String: AnyObject]?, completion: ResultType<AnyObject, APIDataError>.result -> Void)
+}
+
+final class GoogleAPIDataProvider : APIDataProvider {
```
Comment (ochococo) [inline GrandCentralBoard/Utilities/GoogleAPIDataProvider.swift]: `GoogleAPIDataProvider : APIDataProviding` would be better IMO.

File: GrandCentralBoard/Widgets/Watch/Server/GoogleCalendarDataProvider.swift [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+//  GrandCentralBoard
+//
+//  Created by Michał Laskowski on 07.04.2016.
+//  Copyright © 2016 Oktawian Chojnacki. All rights reserved.
+//
+
+import Foundation
+import Result
+
+protocol CalendarDataProvider {
```
Comment (ochococo) [inline GrandCentralBoard/Widgets/Watch/Server/GoogleCalendarDataProvider.swift]: `CalendarDataProviding`?

File: GrandCentralBoard/Widgets/Watch/Server/GoogleCalendarDataProvider.swift [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+    func fetchEventsForCalendar(calendarID: String, completion: (Result<[EventModel], APIDataError>) -> Void) {
+        guard let escapedCalendarID = calendarID.URLEscape() else {
+            assertionFailure("Failure escaping calendar id: \(calendarID)")
+            return
+        }
+        let url = NSURL(string: "https://www.googleapis.com/calendar/v3/calendars/\(escapedCalendarID)/events")!
+        let timeMin = self.dynamicType.dateFormatter.stringFromDate(NSDate())
+        let parameters: [String : AnyObject] = ["maxResults" : 10, "orderyBy" : "startTime",
+                                                "singleEvents" : "true", "maxAttendees" : 1,
```
Comment (ochococo) [inline GrandCentralBoard/Widgets/Watch/Server/GoogleCalendarDataProvider.swift]: `"maxAttendees" : 1` - why?

File: GrandCentralBoard/Widgets/Watch/Server/GoogleCalendarDataProvider.swift [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        formatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ssZ";
+        return formatter
+    }()
+
+    func fetchEventsForCalendar(calendarID: String, completion: (Result<[EventModel], APIDataError>) -> Void) {
+        guard let escapedCalendarID = calendarID.URLEscape() else {
+            assertionFailure("Failure escaping calendar id: \(calendarID)")
+            return
+        }
+        let url = NSURL(string: "https://www.googleapis.com/calendar/v3/calendars/\(escapedCalendarID)/events")!
```
Comment (ochococo) [inline GrandCentralBoard/Widgets/Watch/Server/GoogleCalendarDataProvider.swift]: `https://www.googleapis.com/calendar/v3/calendars/`

Define somewhere as `let`. You have identical literal in line `62`.

File: GrandCentralBoard/Widgets/Watch/Server/GoogleCalendarDataProvider.swift [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+                } catch {
+                    completion(.Failure(.ModelDecodeError(error)))
+                }
+            }
+        }
+    }
+
+    func fetchCalendar(calendarID: String, completion: (Result<CalendarModel, APIDataError>) -> Void) {
+        guard let escapedCalendarID = calendarID.URLEscape() else {
+            assertionFailure("Failure escaping calendar id: \(calendarID)")
```
Comment (ochococo) [inline GrandCentralBoard/Widgets/Watch/Server/GoogleCalendarDataProvider.swift]: May I suggest `completion(.Failure(X))` here?

File: GrandCentralBoardTests/CalendarDataProviderTests.swift [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+        let urlString = url.URLString
+        if urlString.hasSuffix("/events") {
+            completion(.Success(eventsDictionary))
+        } else {
+            completion(.Success(calendarDictionary))
+        }
+    }
+}
+
+class CalendarDataProviderTests : XCTestCase {
```
Comment (ochococo) [inline GrandCentralBoardTests/CalendarDataProviderTests.swift]: 👏 I love this tests!!!

File: GrandCentralBoard/Widgets/Watch/Server/GoogleCalendarDataProvider.swift (hunk 6 above)
Comment (michallaskowski) [inline GrandCentralBoard/Widgets/Watch/Server/GoogleCalendarDataProvider.swift]: I am on it, I was going to add it now but you found it first ;)

File: GrandCentralBoard/Widgets/Watch/Server/GoogleCalendarDataProvider.swift (hunk 4 above)
Comment (michallaskowski) [inline GrandCentralBoard/Widgets/Watch/Server/GoogleCalendarDataProvider.swift]: We do not display attendees, and the minival value here is 1. So to minimize the response size I used the minimal allowed value. I will add a comment about it, ok?

**Comment by ochococo** [comment]
No more comments. Tests passed. @michallaskowski 🎾

**Comment by ochococo** [comment]
screenshot of tower 11-04-2016 13-52-01

Capitalize the first letter in commit message.

**Comment by michallaskowski** [comment]
@ochococo Updated

File: GrandCentralBoard/Utilities/GoogleAPIDataProvider.swift [hunk 8]
```diff
@@ -10,20 +10,26 @@ import Alamofire
...
 
-final class GoogleAPIDataProvider {
+protocol APIDataProviding {
+    func request(method: Method, url: NSURL, parameters: [String: AnyObject]?, completion: ResultType<AnyObject, APIDataError>.result -> Void)
+}
+
+final class GoogleAPIDataProvider : APIDataProviding {
 
-    private let tokenProvider: OAuthTokenProvider
+    private let tokenProvider: OAuthTokenProviding
```
Comment (ochococo) [inline GrandCentralBoard/Utilities/GoogleAPIDataProvider.swift]: `OAuth2TokenProviding` would be more precise.

File: GrandCentralBoard/Widgets/Watch/Server/GoogleCalendarDataProvider.swift [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+    case GetEvents  = "events"
+
+    static private let baseURL = NSURL(string: "https://www.googleapis.com/calendar/v3/calendars/")!
+
+    func URLForCalendar(calendarID: String) -> NSURL? {
+        guard let escapedCalendarID = calendarID.URLEscape() else {
+            assertionFailure("Failure escaping calendar id: \(calendarID)")
+            return nil
+        }
+        return self.dynamicType.baseURL.URLByAppendingPathComponent("\(escapedCalendarID)/\(self.rawValue)")
```
Comment (ochococo) [inline GrandCentralBoard/Widgets/Watch/Server/GoogleCalendarDataProvider.swift]: You could remove `self.dynamicType.`.

Comment (michallaskowski) [inline GrandCentralBoard/Widgets/Watch/Server/GoogleCalendarDataProvider.swift]: It's a static property, I could only replace `self.dynamicType` with `CalendarAPIAction`. I will leave it as it is

File: GrandCentralBoard/Widgets/Watch/Server/GoogleCalendarDataProvider.swift [hunk 10]
```diff
@@ -29,12 +44,14 @@ final class GoogleCalendarDataProvider : CalendarDataProviding {
     }()
 
     func fetchEventsForCalendar(calendarID: String, completion: (Result<[EventModel], APIDataError>) -> Void) {
-        guard let escapedCalendarID = calendarID.URLEscape() else {
-            assertionFailure("Failure escaping calendar id: \(calendarID)")
+                let timeMin = self.dynamicType.dateFormatter.stringFromDate(NSDate())
```
Comment (karolkozub) [inline GrandCentralBoard/Widgets/Watch/Server/GoogleCalendarDataProvider.swift]: What's with the weird indentation?

**Comment by karolkozub** [comment]
I don't have any more comments.

**Comment by michallaskowski** [comment]
@karolkozub Updated

**Pull request merged**